flocker/node/agents/test/test_blockdevice.py

```diff
@@ -10,18 +10,401 @@
+        configuration = Deployment(
+            nodes=frozenset({
+                Node(
+                    hostname=node,
+                    manifestations={dataset_id: manifestation},
+                )
+            })
+        )
+        state = Deployment(nodes=[])
+        api = LoopbackBlockDeviceAPI.from_path(self.mktemp())
+        deployer = BlockDeviceDeployer(
+            hostname=node,
+            block_device_api=api,
+        )
+        changes = deployer.calculate_necessary_state_changes(
+            local_state=NodeState(hostname=node),
+            desired_configuration=configuration,
+            current_cluster_state=state,
+        )
+        mountpoint = deployer.mountroot.child(dataset_id.encode("ascii"))
+        self.assertEqual(
+            InParallel(
+                changes=[
+                    CreateBlockDeviceDataset(
+                        dataset=dataset, mountpoint=mountpoint
+                    )
+                ]),
+            changes
+        )
+
+    def _calculate_changes(self, local_hostname, local_state,
+                           desired_configuration):
+        """
+        Create a ``BlockDeviceDeployer`` and call its
+        ``calculate_necessary_state_changes`` method with the given arguments
+        and an empty cluster state.
+
+        :param unicode local_hostname: The node identifier to give to the
+            ``BlockDeviceDeployer``.
+        :param local_state: As accepted by
+            ``IDeployer.calculate_necessary_state_changes``.
+        :param desired_configuration: As accepted by
+            ``IDeployer.calculate_necessary_state_changes``.
+
+        :return: The return value of
+            ``BlockDeviceDeployer.calculate_necessary_state_changes``.
+        """
+        # Control service still reports that this node has no manifestations.
+        current_cluster_state = Deployment(
+            nodes=frozenset([Node(hostname=local_hostname)])
+        )
+
+        api = LoopbackBlockDeviceAPI.from_path(self.mktemp())
+        deployer = BlockDeviceDeployer(
+            hostname=local_hostname,
+            block_device_api=api,
+        )
+
+        return deployer.calculate_necessary_state_changes(
+            local_state, desired_configuration, current_cluster_state
+        )
+
+    def test_local_state_overrides_cluster_state(self):
+        """
+        ``BlockDeviceDeployer.calculate_necessary_state_changes`` bases its
+        decision about whether it is necessary to create a dataset by
+        inspecting the ``local_state`` argument.  It does this, instead of
+        inspecting the ``current_cluster_state`` argument, because the
+        ``local_state`` is presumed to be the most up-to-date information.
+        ``current_cluster_state`` has traveled a round-trip between the node
+        and the control service and had that period of time to fall behind the
+        actual node state.
+        """
+        local_hostname = u"192.0.2.1"
```

Comment:
I think @exarkun and I discussed this and thought it was useful to have the implementation specific test...but I can't remember our thinking.